proposals/XXXX-extend-property-wrappers-to-function-and-closure-parameters.md

```diff
@@ -0,0 +1,385 @@
+
+let a: A = { @Reference foo in
+  ...
+}
+```
+
+    Becomes:
+
+```swift
+let a: A = { _foo in
+  var foo: Int {
+    get { 
+      _foo.wrappedValue 
+    }
+    set { 
+      _foo.wrappedValue = newValue
+    }
+  }
+
+  var $foo: Int {
+    get { 
+      _foo.projectedValue 
+    }
+  }
+    
+  ...
+}
+```
+
+2. Value Semantics Wrapper
+
+```swift
+@propertyWrapper
+struct Wrapper<Value> {
+  init(wrappedValue: Value) 
+    
+  var wrappedValue: Value {
+    get 
+    set
+  }
+}
+
+typealias B = (inout Wrapper<Int>) -> Void
+
+let b: B = { @Wrapper foo in
+  ...
+}
+```
+
+    Becomes:
+
+```swift
+let b: B = { _foo in
+  var foo: Int {
+    get { 
+      _foo.wrappedValue 
+    }
+    set { 
+      _foo.wrappedValue = newValue
+    }
+    // Since the paramter is marked `inout`
+    // we are allowed to have a mutating setter.
+  }
+  
+  ...
+}
+```
+
+3. Value Semantics Wrapper with Special Initializer
+
+```swift
+@propertyWrapper
+class WrapperObject<Value> {
+  init(wrappedValue: Value) 
+    
+  var wrappedValue: Value {
+    get 
+    set 
+    // Not actually mutating
+    // because `WrapperObject` is
+    // a class.
+  }
+}
+
+typealias C = (WrapperObject<Int>) -> Void
+
+let c: C = { @WrapperObject foo in
+  ...
+}
+```
+
+    Becomes:
+
+```swift
+let c: C = { _foo in
+  var foo: Int {
+    get { 
+      _foo.wrappedValue 
+    }
+    set { 
+      _foo.wrappedValue = newValue
+    }
+    // Since `WrapperObject` has reference
+    // semantics we can include a setter.
+  }
+
+  ...
+}
+```
+
+
+## Source compatibility
+
+This is an additive change with _no_ impact on **source compatibility**.
+
+
+## Effect on ABI stability
+
+This is an additive change with _no_ impact on **ABI stability**.
+
+
+## Effect on API resilience
+
+This is an additive change with _no_ impact on **API resilience**.
+
+
+## Alternatives Considered
+
+### Infer Property Wrappers in Closure Parameters 
+
+TBD
+
+
+### Support `@autoclosure` and `@escaping` in Function Parameters
+
+TBD
+
+
+## Future Directions
+
+### Support property wrapper initialization from a projected value
+
+Today, a property wrapper can be initialized from an instance of its `wrappedValue` type if the wrapper provides a suitable `init(wrappedValue:)`. The same initialization strategy is used in this proposal for property wrapper parameters to allow users to pass a wrapped value as a property wrapper argument. We could extend this model to support initializing a property wrapper from an instance of its `projectedValue` type by allowing property wrappers to define an `init(projectedValue:)` that follows the same rules as `init(wrappedValue:)`. This could allow users to additionally pass a projected value as a property wrapper argument, like so:
+
+```swift
+@propertyWrapper
+struct Wrapper<T> {
+  var wrappedValue: T
+  var projectedValue: ProjectionType
+
+  init(projectedValue: ProjectionType) { ... }
+}
+
+func fn(@Wrapper value: Int) { ... }
```

Comment:
Would you mind if we changed the function name to something like `foo` instead of the abbreviation `fn`?

Reply:
I'm not against changing the example, but I actually am against using `foo` - instead I think we should use a more realistic example throughout the whole document, such as `@Lazy` or `@Clamped`. It would also be nice to use the same example throughout and build upon it to avoid too much context switching for the reader

The reason I'm against using `foo` and `bar` is that "foobar" is derived from some very negative slang.

Reply:
#### [ ... ] realistic example throughout the whole document, such as @Lazy or @Clamped

I agree, I'll use the relevant names when rewriting the examples.

#### The reason I'm against using foo and bar is that "foobar" is derived from some very negative slang.

I had no idea that this was where foo originated from. Thanks for pointing it out!

Reply:
I've changed all the sections up to and including the Proposed Solution. Arguably it's a bit more confusing than using meaningless wrappers; however, the do make the motivation more persuasive. Let me know if there's something you think we could simplify!